File: package/baz/src/i.c

#include <R.h>
#include <R_ext/Lapack.h>

// matrix inverse using cholesky (LAPACK)
// first call DPOTRF and then DPOTRI

void matinv(double *a, int *n, double *result)
{
    int info;
    F77_CALL(dpotrf)("L", n, a, n, &info);
    if (info != 0)
        error("Cholesky decomposition failed");
    F77_CALL(dpotri)("L", n, a, n, &info);
    if (info != 0)
        error("inverse computation failed");
    // but we still only have the lower triangle of the result in a
    int in = n[0];
    for (int i = 0; i < in; i++) {
        result[i + in * i] = a[i + in * i];
        for (int j = 0; j < i; j++) {
            result[i + in * j] = a[i + in * j];
            result[j + in * i] = a[i + in * j];
        }
    }
}

// determinant using cholesky (LAPACK)
// first call DPOTRF and then determinant is square of product
// of diagonal elements of Cholesky factor (L in L L^T)

void matdet(double *a, int *n, double *result)
{
    int info;
    F77_CALL(dpotrf)("L", n, a, n, &info);
    if (info != 0)
        error("Cholesky decomposition failed");
    int in = n[0];
    result[0] = 1.0;
    for (int i = 0; i < in; i++)
        result[0] *= a[i + in * i];
    result[0] *= result[0];
}

// Solves A x = b with the solution x clobbering b (LAPACK)
// first call DPOTRF and then DPOTRS

void matsolve(double *a, double *b, int *nrowb, int *ncolb)
{
    int info;
    F77_CALL(dpotrf)("L", nrowb, a, nrowb, &info);
    if (info != 0)
        error("Cholesky decomposition failed");
    F77_CALL(dpotrs)("L", nrowb, ncolb, a, nrowb, b, nrowb, &info);
    if (info != 0)
        error("solution failed");
}

// Finds x^T A^{- 1} x
// Illustrates use of DPOTRS to avoid matrix inverse

void matsmash(double *a, int *n, double *x, double *result)
{
    int info;
    int ione = 1;
    double *savex = (double *) R_alloc(n[0], sizeof(double));
    F77_CALL(dpotrf)("L", n, a, n, &info);
    if (info != 0)
        error("Cholesky decomposition failed");
    F77_CALL(dcopy)(n, x, &ione, savex, &ione);
    F77_CALL(dpotrs)("L", n, &ione, a, n, x, n, &info);
    if (info != 0)
        error("solution failed");
    result[0] = F77_CALL(ddot)(n, x, &ione, savex, &ione);
}

File: package/baz/R/i.R

matinv <- function(a)
{
    stopifnot(is.numeric(a))
    stopifnot(is.finite(a))
    stopifnot(is.matrix(a))
    stopifnot(a == t(a))
    .C("matinv", a = a, n = nrow(a),
        result = matrix(as.double(0), nrow(a), ncol(a)),
        PACKAGE = "baz")$result
}

matdet <- function(a)
{
    stopifnot(is.numeric(a))
    stopifnot(is.finite(a))
    stopifnot(is.matrix(a))
    stopifnot(a == t(a))
    .C("matdet", a = a, n = nrow(a), result = double(1),
        PACKAGE = "baz")$result
}

matsolve <- function(a, b)
{
    stopifnot(is.numeric(a))
    stopifnot(is.numeric(b))
    stopifnot(is.finite(a))
    stopifnot(is.finite(b))
    stopifnot(is.matrix(a))
    stopifnot(a == t(a))
    b <- as.matrix(b)
    stopifnot(nrow(a) == nrow(b))
    storage.mode(a) <- "double"
    storage.mode(b) <- "double"
    .C("matsolve", a = a, b = b, nrowb = nrow(b), ncolb = ncol(b),
        PACKAGE = "baz")$b
}

matsmash <- function(a, x, fortran = FALSE)
{
    stopifnot(is.numeric(a))
    stopifnot(is.numeric(x))
    stopifnot(is.finite(a))
    stopifnot(is.finite(x))
    stopifnot(is.matrix(a))
    stopifnot(a == t(a))
    stopifnot(nrow(a) == length(x))
    stopifnot(is.logical(fortran))
    stopifnot(length(fortran) == 1)
    if (fortran) {
        .Fortran("footran", a = as.double(a), n = nrow(a), x = as.double(x),
            result = double(1), PACKAGE = "baz")$result
    } else {
        .C("matsmash", a = as.double(a), n = nrow(a), x = as.double(x),
        result = double(1), PACKAGE = "baz")$result
    }
}

